pkg/raft/log.go

```diff
@@ -54,8 +54,28 @@ type raftLog struct {
 	// they will be saved into storage.
 	unstable unstable
 
-	// committed is the highest log position that is known to be in
-	// stable storage on a quorum of nodes.
+	// committed is the highest index in this log known to be committed. This
+	// index logically corresponds to the CommitMark in the coordinate system of
+	// the current raft term: CommitMark{Term: raft.Term, Index: committed}.
+	//
+	// Note that the log's accTerm can be below raft.Term. This means that there
+	// is a "time travel" effect possible when an entry in this log's accTerm
+	// coordinate system is known to be committed before the actual Term at which
+	// it got committed. There is no guarantee that in other logs at terms between
+	// accTerm and raft.Term this entry is the same. It would be unsafe to declare
+	// this entry committed in our accTerm log, because it would have a chance to
+	// be overwritten. But the guarantee below allows it.
+	//
+	// The invariant we rely on is: accTerm <= raft.Term. Raft has given a promise
+	// to never accept log writes from terms < raft.Term. Since the committed
+	// index is in raft.Term commit mark coordinate system, we have the guarantee
```

Comment:
Is this "just" saying that a leader will never tell a follower about a committed index it doesn't already have in the log?

If so: I'm not sure whether "since the committed index is in raft.Term commit mark coordinate system" is a good way to express this. Something being in a coordinate system doesn't guarantee anything, right? It's that we only talk about something in this coordinate system once some conditions are fulfilled. I think the commit index coordinate system is useful, but if we lean on it too much to communicate our invariants, we might be obfuscating rather than explaining.

If not: whatever this means, I couldn't figure it out.